### PR TITLE
Make sure members of type `Class` correctly serialized using AMQP

### PR DESCRIPTION
**Pull request by vkolomeyko**
See: `net.corda.finance.contracts.asset.Cash.Commands.Move` for example use

Also wrote the unit test that exposes the issue(CommandsSerializationTests) and fixed TransactionEncumbranceTests to make it pass in AMQP mode

### CODE REVIEW

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/Schema.kt [hunk 1]
```diff
@@ -377,7 +383,7 @@ private fun fingerprintForType(type: Type, contextType: Type?, alreadySeen: Muta
                 val startingHash = if (isCollectionOrMap(clazz)) {
                     hasher.putUnencodedChars(clazz.name)
                 } else {
-                    fingerprintForObject(type, type, alreadySeen, hasher, factory)
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/Schema.kt]: Why the change here?  Now two objects with the same name but different definitions hash the same.

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/custom/ClassSerializer.kt [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+import net.corda.nodeapi.internal.serialization.amqp.CustomSerializer
+import net.corda.nodeapi.internal.serialization.amqp.SerializerFactory
+
+/**
+ * A serializer for [Class] that uses [ClassProxy] proxy object to write out
+ */
+class ClassSerializer(factory: SerializerFactory) : CustomSerializer.Proxy<Class<*>, ClassSerializer.ClassProxy>(Class::class.java, ClassProxy::class.java, factory) {
+    override fun toProxy(obj: Class<*>): ClassProxy = ClassProxy(obj.name)
+
+    override fun fromProxy(proxy: ClassProxy): Class<*> = Class.forName(proxy.className)
```
Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/custom/ClassSerializer.kt]: Hmmmm..... we need the class loader in `Class.forName`.  `factory.classloader` should be the correct one.

Comment (vkolomeyko) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/custom/ClassSerializer.kt]: Sure - will change

File: node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/Schema.kt (hunk 1 above)
Comment (vkolomeyko) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/Schema.kt]: Well, on this `if` branch we do already know that we are dealing with `ParameterizedType`, therefore using `fingerprintForObject` will be too generic and will fail to explore `ParameterizedType` in full via reflection of properties.
On the other hand `Class` is kind of a singleton, therefore the same logic should apply in this case as we do for `object`s

Comment (vkolomeyko) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/Schema.kt]: I am seeing more problem with this code e.g. SerializationOutputTests - more changes to come in this space

Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/Schema.kt]: But you miss the `alreadySeen` so we can end up in a loop I think.

e.g. 

`data class A(val a: A)` is possible, but loops here.... maybe?

Comment (vkolomeyko) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/Schema.kt]: Code in the latest commit is using:
```
hasher.fingerprintWithCustomSerializerOrElse(factory, clazz, type) {
                        fingerprintForObject(type, type, alreadySeen, hasher, factory)
                    }
```
So, `alreadySeen` and recursion are being catered for, just as before my changes.

Comment (rick-r3) [inline node-api/src/main/kotlin/net/corda/nodeapi/internal/serialization/amqp/Schema.kt]: This look about right to me though....
